connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SourceTaskOffsetCommitter.java

```diff
@@ -67,70 +69,53 @@ public void close(long timeoutMs) {
     public void schedule(final ConnectorTaskId id, final WorkerSourceTask workerTask) {
         synchronized (committers) {
             long commitIntervalMs = config.getLong(WorkerConfig.OFFSET_COMMIT_INTERVAL_MS_CONFIG);
-            ScheduledFuture<?> commitFuture = commitExecutorService.schedule(new Runnable() {
+            ScheduledFuture<?> commitFuture = commitExecutorService.scheduleWithFixedDelay(new Runnable() {
                 @Override
                 public void run() {
                     commit(id, workerTask);
                 }
-            }, commitIntervalMs, TimeUnit.MILLISECONDS);
-            committers.put(id, new ScheduledCommitTask(commitFuture));
+            }, 0, commitIntervalMs, TimeUnit.MILLISECONDS);
+            committers.put(id, commitFuture);
```

Comment:
Several things here: 

1) If the order of the addition to the map (`commiters.put(...)`) needs to happen before the task executes, we just need to reverse the order of calls in `schedule`. That's not influenced by the use of synchronization in both cases. 

2) The `ConcurrentMap` was introduced to make `remove` and `commit` thread-safe in their own respect (you are referring to thread-safety between `commit` and `remove` but I also had in mind thread-safety between two `remove`s for instance)

3) The `synchronized` block in `schedule` is outdated, I agree. `put` needs to happen first, and then scheduling. If the task is removed in the mean time `commit` will return `false`.

So, yes, there was some stale synchronization and wrong order there. I will update the code. Use of concurrent collection should be sufficient.